README.md
# @favoloso/conventional-changelog-emoji

[![Travis CI](https://img.shields.io/travis/com/favoloso/conventional-changelog-emoji.svg)](https://travis-ci.com/favoloso/conventional-changelog-emoji)
[![Coveralls](https://img.shields.io/coveralls/github/favoloso/conventional-changelog-emoji/master.svg)](https://coveralls.io/github/favoloso/conventional-changelog-emoji)
[![npm](https://img.shields.io/npm/v/@favoloso/conventional-changelog-emoji.svg)](https://www.npmjs.com/package/@favoloso/conventional-changelog-emoji)

Conventional Changelog with Emojis support 🎉

## Installation

```sh
yarn add --dev @favoloso/conventional-changelog-emoji
```

or

```sh
npm install --save-dev @favoloso/conventional-changelog-emoji
```

## Usage

```sh
conventional-changelog -p @favoloso/emoji -i CHANGELOG.md -s
```

### Usage with `release-it`

You can use [conventional-changelog
plugin](https://github.com/release-it/conventional-changelog) (v11+) with the
`@favoloso/emoji` preset.

```json
"release-it": {
  "plugins": {
    "@release-it/conventional-changelog": {
      "preset": "@favoloso/emoji",
      "infile": "CHANGELOG.md"
    }
  }
}
```

## Lint (and fix) commit messages

This package provides an additional bin script `emoji-commit-lint`.

The scripts **lints and eventually changes commit messages**, from traditional conventional changelog
format (i.e. `feat: Add a magic feature`) to corresponding emoji (i.e. `✨ Add a magic feature`),
checks and applies correct casing (i.e. `lower-case`) and many more, as configured
by [**Linter Rules**](https://github.com/favoloso/conventional-changelog-emoji/wiki/Linter-Rules).

To use it, install [husky](https://github.com/typicode/husky)

```sh
yarn add --dev husky
```

Now in your package.json add:

```json
{
  "husky": {
    "hooks": {
      "commit-msg": "emoji-commit-lint"
    }
  }
}
```

Now linter will check your commits. Any commit like `<type>: <msg>` will be automatically
transformed with related emoji. See _Available Emojis_ to see available **types**.

## Configuration

The package works as-is, but its behaviour may be customized with the following options.

> _Note:_ This package supports [`cosmiconfig`](https://github.com/davidtheclark/cosmiconfig#readme) to provide configuration options with `favolosoEmoji` module name.

- **`emojis`** (default: `{}`)

  An object allowing you to customize conventional-changelog types used (as _Available Emojis_ table).

  You should provide an object, where the key is the `type` you want to edit (or add), and the value is the
  updated configuration. Configurations will be merged with originals if existing.

  See the [**Custom Emoji** wiki page](https://github.com/favoloso/conventional-changelog-emoji/wiki/Custom-Emojis) for further details.

- **`rules`** (default: `{}`)

  Allows to customize linter rules.
  See the [**Linter Rules** wiki page](https://github.com/favoloso/conventional-changelog-emoji/wiki/Linter-Rules) for further details.

- **`showEmojiPerCommit`** (default: `false`)

  In the changelog, shows emoji for each commit. In the default mode (`false`), emojis are omitted from commits and only the heading contains them (i.e. `🐛 Bug Fixes`)

- **`minorForBreakingInDevelopment`** (default: `true`)

  Breaking changes during development will be considered as _minor_ instead of _major_ bumps (see [semver spec](https://semver.org/#spec-item-4)).

  > 4.  Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

- **`language`** (default: `en`)

  Allows to translate commits group heading in `CHANGELOG.md` (i.e. `🐛 Bug Fixes`) and in linter messages.
  Languages available: `en`, `it`

### Example config in package.json

```json
{
  "favolosoEmoji": {
    "showEmojiPerCommit": false
  }
}
```

## Available Emojis

<!-- prettier-ignore-start -->
<!-- emoji-table -->

 Emoji | Aliases | Type | Type Aliases | Version Bump | In Changelog? | Heading  | Order
 ----- | ------- | ---- | ------------ | ------------ | ------------- | -------  | -----
🚨 |  | `breaking` |  | major | ✅ | 🚨 Breaking Changes | 10
✨ | 🌟, 💫, 🌠 | `feat` |  | minor | ✅ | ✨ Features | 20
🔒 |  | `security` |  | patch | ✅ | 🔒 Security | 25
🛠 |  | `improvement` | `imp` | patch | ✅ | 🛠 Improvements | 30
⚡️ |  | `perf` | `performance` | patch | ✅ | ⚡️ Performance | 35
🐛 | 🐞 | `fix` |  | patch | ✅ | 🐛 Bug Fixes | 40
📚 | 📖 | `docs` | `doc` | patch | ✅ | 📚 Documentation | 50
🏗 | ⚙️ | `chore` | `chores` | patch | ✅ | 🏗 Chore | 60
♻️ |  | `refactor` |  | patch |  | ♻️ Refactoring | 90
🚦 | ✅ | `test` |  | patch |  | 🚦 Test | 90
🎨 | 💄 | `style` | `cleanup` | patch |  | 🎨 Style | 90
📦 |  | `build` | `deps` | patch |  | 📦 Build | 90
🔖 |  | `release` |  | patch |  | 🔖 Release | 90
🚧 |  | `wip` |  | patch |  | 🚧 Wip | 90

<!-- emoji-table -->
<!-- prettier-ignore-end -->


CHANGELOG.md
# [0.10.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.9.0...v0.10.0) (2019-10-24)

### ✨ Features

- Add support for breaking changes bump in commit messages ([1612535](https://github.com/favoloso/conventional-changelog-emoji/commit/1612535))

### 🐛 Bug Fixes

- Replace correctly arguments when linter errors are shown ([79ba400](https://github.com/favoloso/conventional-changelog-emoji/commit/79ba400))

### 📚 Documentation

- Add release-it example ([09207e9](https://github.com/favoloso/conventional-changelog-emoji/commit/09207e9))
- Fix husky snippet ([5fd7d4f](https://github.com/favoloso/conventional-changelog-emoji/commit/5fd7d4f))
- Fixed scripts in the README ([8081f39](https://github.com/favoloso/conventional-changelog-emoji/commit/8081f39))

# [0.9.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.8.0...v0.9.0) (2019-03-02)

### ✨ Features

- Add translations support for headings and rules ([553f3d7](https://github.com/favoloso/conventional-changelog-emoji/commit/553f3d7))

### 🐛 Bug Fixes

- Fix a bug while parsing some emojis (like ⚡️) ([9da7bac](https://github.com/favoloso/conventional-changelog-emoji/commit/9da7bac))

### 📚 Documentation

- Add `language` option to README ([bd219ad](https://github.com/favoloso/conventional-changelog-emoji/commit/bd219ad))

# [0.8.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.7.0...v0.8.0) (2019-03-01)

### ✨ Features

- Add body, header and subject lint rules ([7b7f5a7](https://github.com/favoloso/conventional-changelog-emoji/commit/7b7f5a7))
- Add linter formatted messages ([adfe690](https://github.com/favoloso/conventional-changelog-emoji/commit/adfe690))
- Add linting system ([f669617](https://github.com/favoloso/conventional-changelog-emoji/commit/f669617))

### 🛠 Improvements

- Enable `emoji-require` rule by default ([c1b7587](https://github.com/favoloso/conventional-changelog-emoji/commit/c1b7587))
- Enhanced lint messages formatting ([954b796](https://github.com/favoloso/conventional-changelog-emoji/commit/954b796))

### 🐛 Bug Fixes

- Fix body leading when terminating with \n ([9ed4d17](https://github.com/favoloso/conventional-changelog-emoji/commit/9ed4d17))

### 📚 Documentation

- Adds linter rules documentation ([82bba6c](https://github.com/favoloso/conventional-changelog-emoji/commit/82bba6c))

### 🏗 Chore

- Rename bin to `emoji-commit-lint` ([d271cd3](https://github.com/favoloso/conventional-changelog-emoji/commit/d271cd3))

# [0.7.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.6.0...v0.7.0) (2019-02-27)

### ✨ Features

- Add commit groups sort ([460369b](https://github.com/favoloso/conventional-changelog-emoji/commit/460369b))

### 🏗 Chore

- Add automatic emojis table generation in README ([82eccbd](https://github.com/favoloso/conventional-changelog-emoji/commit/82eccbd))
- Setup prettier ([31ea40c](https://github.com/favoloso/conventional-changelog-emoji/commit/31ea40c))

# [0.6.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.5.0...v0.6.0) (2019-02-26)

### ✨ Features

- Add custom emojis support ([80ab5a8](https://github.com/favoloso/conventional-changelog-emoji/commit/80ab5a8))

### 🛠 Improvements

- Refactor code structure ([62198b8](https://github.com/favoloso/conventional-changelog-emoji/commit/62198b8))

# [0.5.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.4.1...v0.5.0) (2019-02-25)

### ✨ Features

- Add `minorForBreakingInDevelopment` config option ([8824c35](https://github.com/favoloso/conventional-changelog-emoji/commit/8824c35))

### 🏗 Chore

- Add coverage reporting with Codecov ([ca50e15](https://github.com/favoloso/conventional-changelog-emoji/commit/ca50e15))
- Add coveralls ([f590aeb](https://github.com/favoloso/conventional-changelog-emoji/commit/f590aeb))
- Use yarn caching for Travis ([ddf4623](https://github.com/favoloso/conventional-changelog-emoji/commit/ddf4623))

### 📚 Documentation

- Add Coveralls badge ([866a3b0](https://github.com/favoloso/conventional-changelog-emoji/commit/866a3b0))

## [0.4.1](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.4.0...v0.4.1) (2019-02-24)

### 🐛 Bug Fixes

- Fix bug in aliased emoji replacement ([395c15d](https://github.com/favoloso/conventional-changelog-emoji/commit/395c15d))

### 🛠 Improvements

- Enhance list of emojis and aliases ([a62dd58](https://github.com/favoloso/conventional-changelog-emoji/commit/a62dd58))
- Rename cosmiconfig module to "favolosoEmoji" ([e0b8f5b](https://github.com/favoloso/conventional-changelog-emoji/commit/e0b8f5b))

# [0.4.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.3.0...v0.4.0) (2019-02-24)

### ✨ Features

- Add `showEmojiPerCommit` option ([f6b79b5](https://github.com/favoloso/conventional-changelog-emoji/commit/f6b79b5))

### 🏗 Chore

- Add Travis CI ([2a769d8](https://github.com/favoloso/conventional-changelog-emoji/commit/2a769d8))

# [0.3.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.2.0...v0.3.0) (2019-02-24)

### ✨ Features

- ✨`fixAliasedEmoji` config option ([63e5c13](https://github.com/favoloso/conventional-changelog-emoji/commit/63e5c13))

### 🐛 Bug Fixes

- 🐛Fix newline on changelog generation for commit lines ([608c7fc](https://github.com/favoloso/conventional-changelog-emoji/commit/608c7fc))

### 📦 Build

- 📦Edit vscode configuration ([277f8e3](https://github.com/favoloso/conventional-changelog-emoji/commit/277f8e3))
- 📦Increment changelog with ourself ([5e43a0c](https://github.com/favoloso/conventional-changelog-emoji/commit/5e43a0c))

### 🛠 Improvements

- 🛠Enhance changelog templates ([46067c7](https://github.com/favoloso/conventional-changelog-emoji/commit/46067c7))
- ♻️Refactor emoji list ([ab448ca](https://github.com/favoloso/conventional-changelog-emoji/commit/ab448ca))

## [0.2.0](https://github.com/favoloso/conventional-changelog-emoji/compare/v0.1.2...v0.2.0) (2019-02-24)

### ✨ Features


src/emoji/emoji-groups.js
module.exports = [
  {
    emoji: "🐛",
    type: "fix",
    inChangelog: true,
    heading: "🐛 Bug Fixes",
    aliases: ["🐞"],
    index: 40
  },
  {
    emoji: "📚",
    type: "docs",
    inChangelog: true,
    heading: "📚 Documentation",
    aliases: ["📖"],
    typeAliases: ["doc"],
    index: 50
  },
  {
    emoji: "🎨",
    type: "style",
    heading: "🎨 Style",
    aliases: ["💄"],
    typeAliases: ["cleanup"],
    index: 90
  },
  {
    emoji: "♻️",
    type: "refactor",
    heading: "♻️ Refactoring",
    index: 90
  },
  {
    emoji: "🛠",
    type: "improvement",
    inChangelog: true,
    heading: "🛠 Improvements",
    typeAliases: ["imp"],
    index: 30
  },
  {
    emoji: "⚡️",
    type: "perf",
    inChangelog: true,
    heading: "⚡️ Performance",
    typeAliases: ["performance"],
    index: 35
  },
  {
    emoji: "🏗",
    type: "chore",
    inChangelog: true,
    heading: "🏗 Chore",
    aliases: ["⚙️"],
    typeAliases: ["chores"],
    index: 60
  },
  {
    emoji: "✨",
    type: "feat",
    bump: "minor",
    inChangelog: true,
    heading: "✨ Features",
    aliases: ["🌟", "💫", "🌠"],
    index: 20
  },
  {
    emoji: "🚨",
    type: "breaking",
    bump: "major",
    inChangelog: true,
    heading: "🚨 Breaking Changes",
    index: 10
  }, // Non rimuovee
  {
    emoji: "🚦",
    type: "test",
    heading: "🚦 Test",
    aliases: ["✅"],
    index: 90
  },
  {
    emoji: "🔒",
    type: "security",
    inChangelog: true,
    heading: "🔒 Security",
    alisases: ["🔑"],
    index: 25
  },
  {
    emoji: "📦",
    type: "build",
    heading: "📦 Build",
    aliases: [],
    typeAliases: ["deps"],
    index: 90
  },
  {
    emoji: "🔖",
    type: "release",
    index: 90
  },
  { emoji: "🚧", type: "wip", bump: "patch", inChangelog: false, index: 90 }
].sort((g, o) => g.index - o.index);
